### PR TITLE
feat(ast): heritage array is now optional, not empty array

### PR DESCRIPTION
**Pull request by JamesHenry**
I'll probably add the other heritage related alignment points before merging

### CODE REVIEW

**Comment by JamesHenry** [comment]
@armano2 in https://github.com/JamesHenry/typescript-estree/pull/120/commits/e40b9d7c8f1a5718a5a05df15a626cf6a08ff35b I have moved the ignore for abstract classes to node patching.

My gut reaction is that aligning with babel here might not actually a step in the right direction. I feel like this is similar to other things we've had before where it makes the most sense for the construct to have its own unique node type...

What do you think? I could imagine this would be harder to deal with in a linting situation, potentially causing issues with existing rules...

**Comment by armano2** [comment]
i was thinking about this to, if we unify those nodes, there will be no easy way to query them in eslint.

i will just suggest renaming them, ClassImplements to TSClassIMplements (this is no es node)

i will follow name of property from babel:
- extends
- implements

instead of heritage

**Comment by JamesHenry** [comment]
@armano2 I have made all the changes we discussed, please could you take a look?

File: src/convert.ts [hunk 1]
```diff
@@ -2426,13 +2424,15 @@ export default function convert(config: ConvertConfig): ESTreeNode | null {
...
-        heritage: hasImplementsClause
-          ? interfaceHeritageClauses[0].types.map(el =>
-              convertHeritageClause(AST_NODE_TYPES.TSInterfaceHeritage, el)
-            )
-          : []
+        id: convertChild(node.name)
       });
+
+      if (interfaceHeritageClauses.length > 0) {
+        result.extends = interfaceHeritageClauses[0].types.map(el =>
```
Comment (armano2) [inline src/convert.ts]: i have only one concern here

```
interface foo implements bar extends foo {}
```

its going to pick up wrong heritage

Comment (armano2) [reply src/convert.ts]: i know that whas a case before to, but prettier is going to mess up class here

Comment (JamesHenry) [reply src/convert.ts]: Yep good point, I'll fix it

File: src/temp-types-based-on-js-source.ts [hunk 2]
```diff
@@ -34,6 +34,8 @@ export interface ESTreeNode {
   value?: string;
   expression?: ESTreeNode | null;
   decorators?: (ESTreeNode | null)[];
+  implements?: ESTreeNode[];
+  extends?: ESTreeNode[];
```
Comment (armano2) [inline src/temp-types-based-on-js-source.ts]: extends can be `array of nodes` or `node`

extends in class is node, in interface is array

Comment (JamesHenry) [reply src/temp-types-based-on-js-source.ts]: It's superClass in classes right? Extends isn't used?

File: src/convert.ts (hunk 1 above)
Comment (armano2) [reply src/convert.ts]: there is also more broken example
```
interface foo extends bar extends baz {}
```

File: src/temp-types-based-on-js-source.ts (hunk 2 above)
Comment (armano2) [reply src/temp-types-based-on-js-source.ts]: ahh you are right, there was change with that :>

**Comment by armano2** [comment]
typescript also parses code like this correctly:
```
interface foo extends foo.call() {}
```
```
interface foo extends foo() {}
```
```
interface foo extends ({ foo: bar }) {
}
```

**Comment by armano2** [comment]
> One thing is that TSExpressionWithTypeArguments has `expression` not `id`.
> 
> To be honest though, I'm struggling to create valid TypeScript where it is anything other than an `Identifier`. Any idea what expression would be valid in `extends ...`?
> 
> cc @uniqueiniquity

i'm starting think that expression is better suited for this, valid nodes there are `Identifier` and `QualifiedName`

**Comment by JamesHenry** [comment]
Pushing shortly, thanks for the updates, and good spot on the TSQualifiedName

**Comment by JamesHenry** [comment]
@armano2 let me know what you think of the last two commits

File: src/convert.ts [hunk 3]
```diff
@@ -2428,9 +2428,26 @@ export default function convert(config: ConvertConfig): ESTreeNode | null {
...
+          if (heritageClause.token === SyntaxKind.ImplementsKeyword) {
+            result.implements = result.implements || [];
+            result.implements = [
+              ...result.implements,
+              ...heritageClause.types.map(n =>
+                convertHeritageClause(AST_NODE_TYPES.TSInterfaceHeritage, n)
+              )
+            ];
+          }
+        });
```
[the remark this answers is not in the document]
Comment (JamesHenry) [reply src/convert.ts]: 👍

**Comment by JamesHenry** [comment]
Thanks! Let's stabilize for a bit now while we catch the ESLint ecosystem up and take stock

**Comment by armano2** [comment]
all of nodes are now aligned -> most of issues is due to range bugs in babel :)

**Comment by JamesHenry** [comment]
Well apart from all the property and kind tweaks we have to make to nodes in utils right? :)

**Comment by armano2** [comment]
yes

**Pull request merged**